### ISSUE TITLE
Akuli not happy with set_enabled_based_on_tab

### ISSUE THREAD

**Issue by benjamin-kirkbride** [Akuli/porcupine#1396]
Not super happy with this design, but I just want this merged.

_Originally posted by @Akuli in https://github.com/Akuli/porcupine/pull/1349#discussion_r1280652436_
            
Can you expand on this?

**Comment by Akuli** [comment]
It is enough and much simpler to detect whether an action is available or not. We don't need callbacks that run when an action becomes available or not available.

For menubar, I added some sample code to #1343, the idea is to enable/disable menu items when the user opens the menu. For right-click menus, we can similarly update the items to be or not to be grayed out when the menu is opened. For command palette we can check availability as the user is typing and we are about to display matching actions.

**Comment by benjamin-kirkbride** [comment]
> We don't need callbacks that run when an action becomes available or not available.

That's not what that is doing? It's a list of callbacks to **determine** if a _menu item_ is available (which may or may not correlate directly to an action)

**Comment by benjamin-kirkbride** [comment]
That said, triggering all of the enabledness callbacks when you click the menubar makes sense

**Comment by Akuli** [comment]
Ideally all menu items would be actions, and the menubar would just check whether those actions are enabled whenever a user looks into the menubar.

Once we are at that point, we can delete `_menu_item_enabledness_callbacks` and close this issue. I don't think it makes sense to change it now, because it does work.

**Comment by benjamin-kirkbride** [comment]
> Ideally all menu items would be actions, and the menubar would just check whether those actions are enabled whenever a user looks into the menubar.

That is what `_menu_item_enabledness_callbacks` is for :sweat_smile:  unless I'm very confused

**Comment by Moosems** [comment]
Why the flip is the name so long 😂?

**Comment by benjamin-kirkbride** [comment]
> Why the flip is the name so long?

Explicit is better than implicit.

**Comment by rdbende** [comment]
> Why the flip is the name so long 😂?

Oh it's average. Look at this:

https://github.com/Akuli/porcupine/blob/9ff4be00597ec2be3cc7a53f1a383d1f36df46e5/porcupine/settings.py#L969

**Comment by Moosems** [comment]
Simple is better than complex.

**Comment by rdbende** [comment]
It's 2023, autocompletions are a thing.

**Comment by Moosems** [comment]
> Oh it's average. Look at this:
> 
> 
> 
> https://github.com/Akuli/porcupine/blob/9ff4be00597ec2be3cc7a53f1a383d1f36df46e5/porcupine/settings.py#L969

That was a complete sentence 😄.

**Comment by Akuli** [comment]
> > .. menubar would just check whether those actions are enabled whenever a user looks into the menubar.
> 
> That is what `_menu_item_enabledness_callbacks` is for sweat_smile unless I'm very confused

Once all menu items are actions, this can be done in a less fragile way without a global callback list. Each action knows whether it is active or not, and if we have a way to find the action corresponding to each menu item, that's all we need.

**Comment by benjamin-kirkbride** [comment]
> Each action knows whether it is active or not

Not exactly: each action has a list of callbacks that determine whether or not an Action is available or not. This might seem like a pedantic distinction, but the reason I bring it up is because we do need to keep track of what needs to be executed to determine if a menu item is available.

Either way though I think I agree that in time this will become cleaner
